lib/monomorphic.ml

```diff
@@ -0,0 +1,161 @@
+    Format.fprintf fmt "@])"
+#endif
+#ifdef INTLIT
+  | `Intlit x ->
+    Format.fprintf fmt "`Intlit (@[<hov>";
+    Format.fprintf fmt "%S" x;
+    Format.fprintf fmt "@])"
+#endif
+#ifdef FLOAT
+  | `Float x ->
+    Format.fprintf fmt "`Float (@[<hov>";
+    Format.fprintf fmt "%F" x;
+    Format.fprintf fmt "@])"
+#endif
+#ifdef FLOATLIT
+  | `Floatlit x ->
+    Format.fprintf fmt "`Floatlit (@[<hov>";
+    Format.fprintf fmt "%S" x;
+    Format.fprintf fmt "@])"
+#endif
+#ifdef STRING
+  | `String x ->
+    Format.fprintf fmt "`String (@[<hov>";
+    Format.fprintf fmt "%S" x;
+    Format.fprintf fmt "@])"
+#endif
+#ifdef STRINGLIT
+  | `Stringlit x ->
+    Format.fprintf fmt "`Stringlit (@[<hov>";
+    Format.fprintf fmt "%S" x;
+    Format.fprintf fmt "@])"
+#endif
+  | `Assoc xs ->
+    Format.fprintf fmt "`Assoc (@[<hov>";
+    Format.fprintf fmt "@[<2>[";
+    ignore (List.fold_left
+      (fun sep (key, value) ->
+        if sep then
+          Format.fprintf fmt ";@ ";
+          Format.fprintf fmt "(@[";
+          Format.fprintf fmt "%S" key;
+          Format.fprintf fmt ",@ ";
+          pp fmt value;
+          Format.fprintf fmt "@])";
+          true) false xs);
+    Format.fprintf fmt "@,]@]";
+    Format.fprintf fmt "@])"
+  | `List xs ->
+    Format.fprintf fmt "`List (@[<hov>";
+    Format.fprintf fmt "@[<2>[";
+    ignore (List.fold_left
+      (fun sep x ->
+        if sep then
+          Format.fprintf fmt ";@ ";
+          pp fmt x;
+          true) false xs);
+    Format.fprintf fmt "@,]@]";
+    Format.fprintf fmt "@])"
+#ifdef TUPLE
+  | `Tuple tup ->
+    Format.fprintf fmt "`Tuple (@[<hov>";
+    Format.fprintf fmt "@[<2>[";
+    ignore (List.fold_left
+      (fun sep e ->
+         if sep then
+           Format.fprintf fmt ";@ ";
+           pp fmt e;
+           true) false tup);
+    Format.fprintf fmt "@,]@]";
+    Format.fprintf fmt "@])"
+#endif
+#ifdef VARIANT
+  | `Variant (name, value) ->
+    Format.fprintf fmt "`Variant (@[<hov>";
+    Format.fprintf fmt "(@[";
+    Format.fprintf fmt "%S" name;
+    Format.fprintf fmt ",@ ";
+    (match value with
+      | None -> Format.pp_print_string fmt "None"
+      | Some x ->
+        Format.pp_print_string fmt "(Some ";
+        pp fmt x;
+        Format.pp_print_string fmt ")");
+    Format.fprintf fmt "@])";
+    Format.fprintf fmt "@])"
+#endif
+
+let show x =
+  Format.asprintf "%a" pp x
+
+let rec equal a b =
+  match a, b with
+  | `Null, `Null -> true
+  | `Bool a, `Bool b -> a = b
+#ifdef INT
+  | `Int a, `Int b -> a = b
+#endif
+#ifdef INTLIT
+    | `Intlit a, `Intlit b -> a = b
+#endif
+#ifdef FLOAT
+    | `Float a, `Float b -> a = b
+#endif
+#ifdef FLOATLIT
+    | `Floatlit a, `Floatlit b -> a = b
+#endif
+#ifdef STRING
+    | `String a, `String b -> a = b
+#endif
+#ifdef STRINGLIT
+    | `Stringlit a, `Stringlit b -> a = b
+#endif
+    | `Assoc xs, `Assoc ys ->
+      let xs, ys = match sort (`Assoc xs), sort (`Assoc ys) with
```

Comment:
That's a good point. The issue I see here is that just sorting that way will cause problems with duplicate keys, since `{"a": 42, "a": 23}` might or might not be equal depending on whether it will reorder the duplicate pair in the alist. And sorting "properly" requires to sort by key-first value-second, but for on the value we'd need `compare`.

Maybe I should just bite the bullet and implement compare and then sort in a sensible way, what do you think?

Reply:
Although Yojson do accept such values, given what the [JSON RFC](https://tools.ietf.org/html/rfc8259#section-4) specifies I think we shouldn't bother with key duplicates and go with a sort on keys solely, cf. my suggestion on the doc comment below.

Because `compare` will rely on a sort itself anyway I don't think implementing it will really help figure this out. We will just migrate this issue from being `equal`s problem to being `compare`'s one. It also feels weird to have `equal` rely on `compare` in any other way than just using it straight and comparing the result to `0`.

I'd say let's not support duplicate in `equal` for now

Reply:
Okay, I implemented it with `stable_sort` which I think is a decent compromise between practicality, performance and correctness.